aten/src/ATen/native/Gesv.h

```diff
@@ -1,30 +1,94 @@
+#include <utility>
 #include "ATen/ATen.h"
 
 namespace at { namespace native {
 
-static inline void checkInputs(const Tensor& self, const Tensor& A) {
-  if (A.size(-1) != A.size(-2)) {
-    AT_ERROR("A must be batches of square matrices, "
-        "but they are %lld by %lld matrices",
-        (long long)A.size(-1), (long long)A.size(-2));
+static inline bool isTransposeContiguous(Tensor& self) {
+  return self.dim() == 2 &&
+          self.stride(0) == 1 &&
+          self.stride(1) == self.size(0);
+}
+
+/* gesv takes (self, A) and returns (sol, lu).
+ * (i)  output tensors (sol, lu) may be same as input tensors (self, A)
+ * (ii) for 2D matrices, .t_() represents their column-major format
+ *
+ * Before passing pointers to Lapack, we need to ensure that these pointers
+ * represent Fortran-contiguous tensors in column-major format
+ *
+ * Cases:
+ * 1) `out` has correct shape but elements do not form a contiguous
+ * chunk of memory. Since shape is correct, we don't resize_ it. Instead, we
+ * clone the input tensor into a buffer, use the buffer for Lapack and finally
+ * copy the buffer to the output tensor.
+ *
+ * 2) out.t() is contiguous:
+ *    (i)  &in == &out: use out.data() as is. Do nothing
+ *    (ii) &in != &out: copy in.t() to out.t()
+ * 3) out.t() is not contiguous:
+ *    - resize_ should fix contiguity/size issues
+ *    (i)  &in == &out: copy in.t().clone() to out (same tensor)
+ *    (ii) &in != &out: copy in.t() to out
+ */
+static inline void prepareTensorsForLapack(
+    const Tensor& in, Tensor& out, Tensor& temp) {
+  int64_t x = in.size(0);
+  int64_t y = (in.dim() == 1) ? 1 : in.size(1);
+  bool out_tc = isTransposeContiguous(out);
+  bool out_correct_shape =
+    out.dim() == 2 && out.size(0) == x && out.size(1) == y;
+
+  // view potential 1D `in` as 2D
+  auto in_t = in.view({x, y}).t_();
+
+  if (!out_tc && !out.is_contiguous() && out_correct_shape) {
```

Comment:
There are a lot of conditions mixed together here so it's still hard for me to tell what's going on in each case. i.e., in the `if (!out_tc)` case, what other booleans hold? Since out is being resized, I think that `out_correct_shape` is False, but I'm not entirely sure.

Could you change the structure of the if statements to something like the following (pseudocode)
```
if (out_correct_shape):
    if !out_tc and !out.is_contiguous():
       ...
else:
    if (out_tc):
        ...
    else:
        ...
```
I think that is what you're trying to express, please let me know if not otherwise

Reply:
Well, not exactly, since we still want to resize when it is not the correct shape, it's just that the `resize` in the latter two conditions doesn't do anything if `out_correct_shape` is true. I think this is probably the most succinct way to write it without making the if-else ladder long and unnecessarily nested (since the exact cases are described in the comment above, it shouldn't be too hard to understand)

Reply:
I'm having a hard time understanding why the cases are mutually exclusive (are they?). It'll probably be more clear if I wrote down a truth table, but at some point this exceeds my threshold for complexity...